File: apps/desktop/src/api/tags.ts
/**
 * Tags API client for GManager Desktop
 *
 * This module provides TypeScript functions that call the Tauri Rust backend
 * for tag management operations.
 *
 * @module api/tags
 */

import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Tag returned from the API
 */
export interface ApiTag {
  /** Unique identifier for the tag */
  id: string;
  /** Tag name */
  name: string;
  /** Tag color (hex) */
  color: string;
  /** ISO timestamp when the tag was created */
  created_at: string;
}

/**
 * Payload for creating a new tag
 */
export interface CreateTagPayload {
  /** Tag name */
  name: string;
  /** Tag color (hex), defaults to '#10b981' */
  color?: string;
}

/**
 * Payload for updating an existing tag
 */
export interface UpdateTagPayload {
  /** The tag ID to update */
  id: string;
  /** New tag name (optional) */
  name?: string;
  /** New tag color (optional) */
  color?: string;
}

/**
 * Payload for adding a tag to an account
 */
export interface AddTagToAccountPayload {
  /** The account ID */
  account_id: string;
  /** The tag ID */
  tag_id: string;
}

// ============================================================================
// Error Handling
// ============================================================================

/**
 * Error type for tag API operations
 */
export class TagApiError extends Error {
  constructor(
    message: string,
    public code?: string,
    public originalError?: unknown
  ) {
    super(message);
    this.name = 'TagApiError';
  }
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get all tags
 *
 * Returns all tags ordered by name.
 *
 * @returns Array of all tags
 * @throws TagApiError if database error occurs
 *
 * @example
 * ```ts
 * const tags = await getTags();
 * console.log(`Found ${tags.length} tags`);
 * ```
 */
export async function getTags(): Promise<ApiTag[]> {
  try {
    return await invoke<ApiTag[]>('get_tags_command');
  } catch (error) {
    throw new TagApiError(
      'Failed to get tags',
      'GET_TAGS_ERROR',
      error
    );
  }
}

/**
 * Get a single tag by ID
 *
 * @param id - The tag ID
 * @returns The tag
 * @throws TagApiError if tag not found
 *
 * @example
 * ```ts
 * const tag = await getTag('1');
 * console.log(tag.name);
 * ```
 */
export async function getTag(id: string): Promise<ApiTag> {
  try {
    return await invoke<ApiTag>('get_tag_command', { id });
  } catch (error) {
    throw new TagApiError(
      `Failed to get tag: ${id}`,
      'GET_TAG_ERROR',
      error
    );
  }
}

/**
 * Create a new tag
 *
 * @param tag - The tag data to create
 * @returns The ID of the created tag
 * @throws TagApiError if validation fails or name already exists
 *
 * @example
 * ```ts
 * const id = await createTag({
 *   name: 'important',
 *   color: '#ef4444',
 * });
 * ```
 */
export async function createTag(tag: CreateTagPayload): Promise<string> {
  try {
    return await invoke<string>('create_tag_command', { tag });
  } catch (error) {
    throw new TagApiError(
      'Failed to create tag',
      'CREATE_ERROR',
      error
    );
  }
}

/**
 * Update an existing tag
 *
 * Only include the fields that should be updated.
 *
 * @param tag - The tag updates with ID
 * @throws TagApiError if tag not found or name already exists
 *
 * @example
 * ```ts
 * await updateTag({
 *   id: '1',
 *   name: 'critical',
 *   color: '#dc2626',
 * });
 * ```
 */
export async function updateTag(tag: UpdateTagPayload): Promise<void> {
  try {
    await invoke('update_tag_command', { tag });
  } catch (error) {
    throw new TagApiError(
      `Failed to update tag: ${tag.id}`,
      'UPDATE_ERROR',
      error
    );
  }
}

/**
 * Delete a tag by ID
 *
 * Note: The tag will be removed from all accounts that have it.
 *
 * @param id - The tag ID to delete
 * @throws TagApiError if tag not found
 *
 * @example
 * ```ts
 * await deleteTag('1');
 * ```
 */
export async function deleteTag(id: string): Promise<void> {
  try {
    await invoke('delete_tag_command', { id });
  } catch (error) {
    throw new TagApiError(
      `Failed to delete tag: ${id}`,
      'DELETE_ERROR',
      error
    );
  }
}

/**
 * Add a tag to an account
 *
 * Creates an association between an account and a tag.
 * If the association already exists, this is a no-op.
 *
 * @param accountId - The account ID
 * @param tagId - The tag ID
 * @throws TagApiError if account or tag not found
 *
 * @example
 * ```ts
 * await addTagToAccount('account-123', 'tag-456');
 * ```
 */
export async function addTagToAccount(
  accountId: string,
  tagId: string
): Promise<void> {
  try {
    await invoke('add_tag_to_account_command', {
      account_id: accountId,
      tag_id: tagId,
    });
  } catch (error) {
    throw new TagApiError(
      `Failed to add tag ${tagId} to account ${accountId}`,
      'ADD_TAG_ERROR',
      error
    );
  }
}

/**
 * Remove a tag from an account
 *
 * Deletes the association between an account and a tag.
 *
 * @param accountId - The account ID
 * @param tagId - The tag ID
 * @throws TagApiError if association not found
 *
 * @example
 * ```ts
 * await removeTagFromAccount('account-123', 'tag-456');
 * ```
 */
export async function removeTagFromAccount(
  accountId: string,
  tagId: string
): Promise<void> {
  try {
    await invoke('remove_tag_from_account_command', {
      account_id: accountId,
      tag_id: tagId,
    });
  } catch (error) {
    throw new TagApiError(
      `Failed to remove tag ${tagId} from account ${accountId}`,
      'REMOVE_TAG_ERROR',
      error
    );
  }
}

/**
 * Get all tags for an account
 *
 * Returns all tags associated with the specified account.
 *
 * @param accountId - The account ID
 * @returns Array of tags associated with the account
 * @throws TagApiError if account not found
 *
 * @example
 * ```ts
 * const tags = await getAccountTags('account-123');
 * console.log(`Account has ${tags.length} tags`);
 * ```
 */
export async function getAccountTags(accountId: string): Promise<ApiTag[]> {
  try {
    return await invoke<ApiTag[]>('get_account_tags_command', {
      account_id: accountId,
    });
  } catch (error) {
    throw new TagApiError(
      `Failed to get tags for account: ${accountId}`,
      'GET_ACCOUNT_TAGS_ERROR',
      error
    );
  }
}

/**
 * Set multiple tags for an account
 *
 * Replaces all existing tags on an account with the provided list.
 * Tags not in the list will be removed.
 *
 * @param accountId - The account ID
 * @param tagIds - Array of tag IDs to associate with the account
 * @throws TagApiError if operation fails
 *
 * @example
 * ```ts
 * await setAccountTags('account-123', ['tag-1', 'tag-2', 'tag-3']);
 * ```
 */
export async function setAccountTags(
  accountId: string,
  tagIds: string[]
): Promise<void> {
  try {
    await invoke('set_account_tags_command', {
      account_id: accountId,
      tag_ids: tagIds,
    });
  } catch (error) {
    throw new TagApiError(
      `Failed to set tags for account: ${accountId}`,
      'SET_TAGS_ERROR',
      error
    );
  }
}

/**
 * Get the number of accounts that have a specific tag
 *
 * @param tagId - The tag ID
 * @returns The number of accounts with this tag
 * @throws TagApiError if database error occurs
 *
 * @example
 * ```ts
 * const count = await getTagAccountsCount('tag-1');
 * console.log(`Accounts with tag: ${count}`);
 * ```
 */
export async function getTagAccountsCount(tagId: string): Promise<number> {
  try {
    return await invoke<number>('get_tag_accounts_count_command', {
      tag_id: tagId,
    });
  } catch (error) {
    throw new TagApiError(
      `Failed to get account count for tag: ${tagId}`,
      'GET_COUNT_ERROR',
      error
    );
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Convert an API tag to a shared Tag type
 * Useful for compatibility with other parts of the application
 *
 * @param apiTag - The tag from the API
 * @returns A Tag matching the shared types
 */
export function apiTagToTag(apiTag: ApiTag): {
  id: string;
  name: string;
  color: string;
  createdAt: string;
} {
  return {
    id: apiTag.id,
    name: apiTag.name,
    color: apiTag.color,
    createdAt: apiTag.created_at,
  };
}

/**
 * Default color palette for tags
 */
export const DEFAULT_TAG_COLORS = [
  '#10b981', // Emerald
  '#3b82f6', // Blue
  '#f59e0b', // Amber
  '#ef4444', // Red
  '#8b5cf6', // Violet
  '#ec4899', // Pink
  '#14b8a6', // Teal
  '#f97316', // Orange
  '#6366f1', // Indigo
  '#84cc16', // Lime
];

/**
 * Generate a default color for a new tag
 *
 * @param index - Index to use for color selection
 * @returns A hex color string
 */
export function getDefaultTagColor(index: number = 0): string {
  return DEFAULT_TAG_COLORS[index % DEFAULT_TAG_COLORS.length];
}

/**
 * Suggested tag names for common use cases
 */
export const SUGGESTED_TAG_NAMES = [
  'important',
  'work',
  'personal',
  'social',
  'finance',
  'shopping',
  'entertainment',
  'development',
  'production',
  'testing',
];

/**
 * Get a suggested tag name based on common patterns
 *
 * @param index - Index to use for name selection
 * @returns A suggested tag name
 */
export function getSuggestedTagName(index: number = 0): string {
  return SUGGESTED_TAG_NAMES[index % SUGGESTED_TAG_NAMES.length];
}

File: apps/desktop/src-tauri/src/tags.rs
//! Tag CRUD operations for GManager Desktop
//!
//! This module provides Tauri commands for tag management.
//!
//! # Tag Management
//!
//! Tags are used to categorize accounts with flexible labeling.
//! Each account can have multiple tags attached to it.

use crate::db::{CreateTag, Database, DbError, Tag};

// ============================================================================
// Type Definitions for Frontend/Backend Communication
// ============================================================================

/// Tag representation sent to frontend
#[derive(Debug, Clone, serde::Serialize)]
pub struct TagDto {
    pub id: String,
    pub name: String,
    pub color: String,
    pub created_at: String,
}

/// Tag creation payload from frontend
#[derive(Debug, Clone, serde::Deserialize)]
pub struct CreateTagPayload {
    pub name: String,
    #[serde(default)]
    pub color: Option<String>,
}

/// Tag update payload from frontend
#[derive(Debug, Clone, serde::Deserialize)]
pub struct UpdateTagPayload {
    pub id: String,
    pub name: Option<String>,
    pub color: Option<String>,
}

/// Payload for setting multiple tags on an account
#[derive(Debug, Clone, serde::Deserialize)]
pub struct SetAccountTagsPayload {
    pub account_id: String,
    pub tag_ids: Vec<String>,
}

// ============================================================================
// Error Types
// ============================================================================

/// Tag operations error type
#[derive(Debug, thiserror::Error)]
pub enum TagError {
    /// Database error
    #[error("Database error: {0}")]
    Database(#[from] DbError),

    /// Invalid input
    #[error("Invalid input: {0}")]
    InvalidInput(String),

    /// Tag not found
    #[error("Tag not found: {0}")]
    NotFound(String),
}

/// Result type for tag operations
pub type TagResult<T> = std::result::Result<T, TagError>;

// ============================================================================
// Conversion Functions
// ============================================================================

/// Convert database tag to frontend tag
fn db_to_frontend_tag(db_tag: Tag) -> TagDto {
    TagDto {
        id: db_tag.id.to_string(),
        name: db_tag.name,
        color: db_tag.color,
        created_at: db_tag.created_at,
    }
}

/// Convert frontend creation payload to database creation payload
fn frontend_to_db_create(payload: CreateTagPayload) -> TagResult<CreateTag> {
    // Validate required fields
    let name = payload.name.trim();
    if name.is_empty() {
        return Err(TagError::InvalidInput("Tag name is required".to_string()));
    }

    // Validate color format (basic hex check)
    let color = payload.color.unwrap_or_else(|| "#10b981".to_string());
    if !color.starts_with('#') || (color.len() != 7 && color.len() != 4) {
        return Err(TagError::InvalidInput(
            "Color must be a valid hex color (e.g., #10b981)".to_string()
        ));
    }

    Ok(CreateTag {
        name: name.to_string(),
        color,
    })
}

// ============================================================================
// Tauri Commands
// ============================================================================

/// Get all tags
///
/// Returns all tags ordered by name.
#[tauri::command]
pub fn get_tags_command(
    db: tauri::State<Database>,
) -> Result<Vec<TagDto>, String> {
    let db_tags = db.get_tags()
        .map_err(|e| e.to_string())?;

    Ok(db_tags
        .into_iter()
        .map(db_to_frontend_tag)
        .collect())
}

/// Get a single tag by ID
#[tauri::command]
pub fn get_tag_command(
    db: tauri::State<Database>,
    id: String,
) -> Result<TagDto, String> {
    let tag_id = id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    let db_tag = db.get_tag(tag_id)
        .map_err(|e| e.to_string())?;

    Ok(db_to_frontend_tag(db_tag))
}

/// Create a new tag
#[tauri::command]
pub fn create_tag_command(
    db: tauri::State<Database>,
    tag: CreateTagPayload,
) -> Result<String, String> {
    let create_tag = frontend_to_db_create(tag)
        .map_err(|e| e.to_string())?;

    let id = db.create_tag(create_tag)
        .map_err(|e| e.to_string())?;

    Ok(id.to_string())
}

/// Update an existing tag
///
/// Only updates provided fields.
#[tauri::command]
pub fn update_tag_command(
    db: tauri::State<Database>,
    tag: UpdateTagPayload,
) -> Result<(), String> {
    let tag_id = tag.id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    // Get existing tag to fill in unchanged fields
    let existing = db.get_tag(tag_id)
        .map_err(|e| e.to_string())?;

    // Build update payload with new values or existing ones
    let update_payload = CreateTag {
        name: tag.name.unwrap_or_else(|| existing.name),
        color: tag.color.unwrap_or_else(|| existing.color),
    };

    db.update_tag(tag_id, update_payload)
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Delete a tag by ID
///
/// Note: The tag will be removed from all accounts that have it.
#[tauri::command]
pub fn delete_tag_command(
    db: tauri::State<Database>,
    id: String,
) -> Result<(), String> {
    let tag_id = id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    db.delete_tag(tag_id)
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Add a tag to an account
///
/// Creates an association between an account and a tag.
/// If the association already exists, this is a no-op.
#[tauri::command]
pub fn add_tag_to_account_command(
    db: tauri::State<Database>,
    account_id: String,
    tag_id: String,
) -> Result<(), String> {
    let aid = account_id.parse::<i64>()
        .map_err(|_| "Invalid account ID".to_string())?;

    let tid = tag_id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    db.add_tag_to_account(aid, tid)
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Remove a tag from an account
///
/// Deletes the association between an account and a tag.
#[tauri::command]
pub fn remove_tag_from_account_command(
    db: tauri::State<Database>,
    account_id: String,
    tag_id: String,
) -> Result<(), String> {
    let aid = account_id.parse::<i64>()
        .map_err(|_| "Invalid account ID".to_string())?;

    let tid = tag_id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    db.remove_tag_from_account(aid, tid)
        .map_err(|e| e.to_string())?;

    Ok(())
}

/// Get all tags for an account
///
/// Returns all tags associated with the specified account.
#[tauri::command]
pub fn get_account_tags_command(
    db: tauri::State<Database>,
    account_id: String,
) -> Result<Vec<TagDto>, String> {
    let aid = account_id.parse::<i64>()
        .map_err(|_| "Invalid account ID".to_string())?;

    let db_tags = db.get_account_tags(aid)
        .map_err(|e| e.to_string())?;

    Ok(db_tags
        .into_iter()
        .map(db_to_frontend_tag)
        .collect())
}

/// Set multiple tags for an account
///
/// Replaces all existing tags on an account with the provided list.
/// Tags not in the list will be removed.
#[tauri::command]
pub fn set_account_tags_command(
    db: tauri::State<Database>,
    payload: SetAccountTagsPayload,
) -> Result<(), String> {
    let account_id = payload.account_id.parse::<i64>()
        .map_err(|_| "Invalid account ID".to_string())?;

    // Parse all tag IDs
    let mut tag_ids = Vec::new();
    for tag_id_str in payload.tag_ids {
        let tid = tag_id_str.parse::<i64>()
            .map_err(|_| "Invalid tag ID".to_string())?;
        tag_ids.push(tid);
    }

    // Get current tags for the account
    let current_tags = db.get_account_tags(account_id)
        .map_err(|e| e.to_string())?;

    // Remove tags that are not in the new list
    for current_tag in &current_tags {
        if !tag_ids.contains(&current_tag.id) {
            db.remove_tag_from_account(account_id, current_tag.id)
                .map_err(|e| e.to_string())?;
        }
    }

    // Add tags that are in the new list but not currently associated
    for tag_id in tag_ids {
        // Check if already associated
        let is_associated = current_tags.iter().any(|t| t.id == tag_id);
        if !is_associated {
            db.add_tag_to_account(account_id, tag_id)
                .map_err(|e| e.to_string())?;
        }
    }

    Ok(())
}

/// Get the number of accounts that have a specific tag
#[tauri::command]
pub fn get_tag_accounts_count_command(
    db: tauri::State<Database>,
    tag_id: String,
) -> Result<i64, String> {
    let tid = tag_id.parse::<i64>()
        .map_err(|_| "Invalid tag ID".to_string())?;

    db.get_tag_accounts_count(tid)
        .map_err(|e| e.to_string())
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_frontend_to_db_create_valid() {
        let payload = CreateTagPayload {
            name: "important".to_string(),
            color: Some("#ef4444".to_string()),
        };

        let result = frontend_to_db_create(payload).unwrap();

        assert_eq!(result.name, "important");
        assert_eq!(result.color, "#ef4444");
    }

    #[test]
    fn test_frontend_to_db_create_defaults() {
        let payload = CreateTagPayload {
            name: "work".to_string(),
            color: None,
        };

        let result = frontend_to_db_create(payload).unwrap();

        assert_eq!(result.name, "work");
        assert_eq!(result.color, "#10b981");
    }

    #[test]
    fn test_frontend_to_db_create_empty_name_fails() {
        let payload = CreateTagPayload {
            name: "".to_string(),
            color: None,
        };

        let result = frontend_to_db_create(payload);
        assert!(result.is_err());
    }

    #[test]
    fn test_frontend_to_db_create_whitespace_name_fails() {
        let payload = CreateTagPayload {
            name: "   ".to_string(),
            color: None,
        };

        let result = frontend_to_db_create(payload);
        assert!(result.is_err());
    }

    #[test]
    fn test_frontend_to_db_create_invalid_color_fails() {
        let payload = CreateTagPayload {
            name: "test".to_string(),
            color: Some("invalid".to_string()),
        };

        let result = frontend_to_db_create(payload);
        assert!(result.is_err());
    }

    #[test]
    fn test_db_to_frontend_conversion() {
        let db_tag = Tag {
            id: 456,
            name: "critical".to_string(),
            color: "#dc2626".to_string(),
            created_at: "2024-01-01T00:00:00Z".to_string(),
        };

        let result = db_to_frontend_tag(db_tag);

        assert_eq!(result.id, "456");
        assert_eq!(result.name, "critical");
        assert_eq!(result.color, "#dc2626");
        assert_eq!(result.created_at, "2024-01-01T00:00:00Z");
    }
}
